Use ccache in GHA pipelines

jonashaag opened the pull request:
Saves around 5–10 minutes in each of the GHA builds.

Also cleans up some duplicated code.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

pep8speaks commented:
Hello @jonashaag! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`setup.py`](https://github.com/pandas-dev/pandas/blob/39761918c88438da9d7282bc99c2cbed75c9f36c/setup.py):

> [Line 228:9](https://github.com/pandas-dev/pandas/blob/39761918c88438da9d7282bc99c2cbed75c9f36c/setup.py#L228): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 559:5](https://github.com/pandas-dev/pandas/blob/39761918c88438da9d7282bc99c2cbed75c9f36c/setup.py#L559): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

mroeschke commented on .github/actions/setup-ccache/action.yml:
```diff
@@ -0,0 +1,17 @@
+      shell: bash
+
+    - name: Setup ccache
+      uses: hendrikmuhs/ccache-action@2181be813387616fc2b8dca72d3ff8b912b25f73
```
Can you reference a version instead of a git hash?

jonashaag replied:
Did this for security purposes but I'm happy to change it if you're not concerned about that.

mroeschke replied:
Shouldn't a v1.x be stable?

jonashaag replied:
I was worried more about security than stability. Tags can be force pushed with malicious code.

mroeschke replied:
Should we not trust this particular action author?

I understand the security concern, but since we also reference tags in other actions, to potentially pick up bug fixes/performance improvements), IMO that's a worthy tradeoff. Open to other core team opinion

mroeschke commented on .github/actions/setup-ccache/action.yml:
```diff
@@ -0,0 +1,17 @@
+    - name: Setup ccache
+      uses: hendrikmuhs/ccache-action@2181be813387616fc2b8dca72d3ff8b912b25f73
+      with:
+        key: ${{ runner.os }}--${{ runner.arch }}--${{ github.workflow }}--${{ steps.get-date.outputs.today }}--${{ inputs.extra-cache-key }}
```
So the cache won't be used if a pyx, .c, etc file is modified?

jonashaag replied:
The cache will always be used but if a pyx (=> c) file is modified the cache for that particular object will not be used. You can think of ccache as something along the lines of

```
cache_key = compute_hash(preprocess_file(myfile))
if cache_key not in cache:
    cache[cache_key] = run_compiler()
return cache[cache_key]
```

mroeschke replied:
Right, could you explain how the hash is being computed here based on a push?

jonashaag replied:
Ccache computes it from the preprocessed C source code.

jonashaag replied:
All we have to deal with here is syncing the ccache folder to/from GHA cache storage. This is what we do above with the key. We can never have bad cache invalidation in the sense that we use stale compilation results (unless there is a bug in Ccache)

mroeschke commented on .github/workflows/code-checks.yml:
```diff
@@ -76,29 +66,20 @@ jobs:
-
     - name: Run checks on imported code
       run: ci/code_checks.sh code
-      if: ${{ steps.build.outcome == 'success' }}
```
I think we would still want this to keep these checks if pandas cant be set up?

jonashaag replied:
I'm really confused about GHA's steps execution behaviour. My take was that those are redundant because GHA will stop executing the job upon the first failed step. Do you know where I can find documentation on that topic? I haven't found anything useful.

mroeschke commented on .github/workflows/code-checks.yml:
```diff
@@ -148,7 +116,6 @@ jobs:
         if grep "failed" benchmarks.log > /dev/null ; then
             exit 1
         fi
-      if: ${{ steps.build.outcome == 'success' }}
```
Same here

mroeschke commented on .github/workflows/datamanager.yml:
```diff
@@ -13,13 +13,20 @@ on:
 env:
-  ENV_FILE: environment.yml
   PANDAS_CI: 1
+  PYTEST_WORKERS: "auto"
```
The name of this file changed.

jonashaag replied:
I'm happy to revert that typo fix if you don't like it.

mroeschke replied:
Ah right looks like the original was a typo. Looks fine as is

jbrockmendel commented:
@jonashaag can you merge main

jonashaag commented:
ping @mroeschke

jonashaag commented:
Trying Sccache right now because it has support for Windows/Visual Studio compiler

jonashaag commented:
I can't figure out how to tell `build_ext` to use a different compiler for Windows. It always seems to use `cl.exe`. @mroeschke @jbrockmendel any ideas?

jonashaag commented:
Awesome, this seems to work great.

jonashaag committed:
Add ccache to GHA builds

- Refactor duplicated code from GHA workflows to the 'build-pandas' and
  'setup' actions.
- Use Mamba over Conda everywhere.
- Add sccache using 'hendrikmuhs/ccache-action'.
- General cleanup of GHA workflows.

jonashaag commented:
Awesome, this also works for Windows now. Not yet implemented for Azure pipelines though, so this only applies to the `python-dev` Windows builds. Would like to test this for a few days and then port it to Azure pipelines as well.

mroeschke commented on .github/workflows/code-checks.yml:
```diff
@@ -94,11 +82,9 @@ jobs:
 
     - name: Run typing validation
       run: ci/code_checks.sh typing
-      if: ${{ steps.build.outcome == 'success' }}
```
I think these 2 `if`s should be here

jonashaag replied:
I don't think so because both require a working `pandas` import, or am I missing something?

mroeschke replied:
In my experience these steps will still "run" and fail for the reason you brought up. i.e. it reduces the noise in the build outputs by just skipping this is the build failed

jonashaag replied:
Oh you are right we actually wanted the same thing but I confused the meaning of this option. Will change

mroeschke commented on .github/workflows/code-checks.yml:
```diff
@@ -148,7 +121,6 @@ jobs:
         if grep "failed" benchmarks.log > /dev/null ; then
             exit 1
         fi
-      if: ${{ steps.build.outcome == 'success' }}
```
Same this should still be here I think

mroeschke commented:
Looks pretty good! Have you observed pushing a commit a day later and seeing the ccache being rebuilt?

jonashaag commented:
I have observed that numerous times while developing this but we can also retrigger this specific version of the PR tomorrow and have a look.

jonashaag commented:
There is one important caveat with this PR that I forgot to mention, which is that it currently uses an unmerged version of `ccache-action` (https://github.com/hendrikmuhs/ccache-action/pull/47)

jreback commented:
can you rebase again (as merged the mamba change).

jonashaag committed:
Switch order of name: to be less confusing

jonashaag committed:
Fix continue condition

jonashaag commented:
I'm working on an extension of this that also uses Micromamba for env setup (where I recently implemented caching of entire Conda envs). It removes another 5 minutes from the builds. https://github.com/pandas-dev/pandas/pull/46346

jonashaag commented:
Example before, almost 10 minutes for Conda setup + Pandas build
<img width="95" alt="Screenshot 2022-03-12 at 22 46 43" src="https://user-images.githubusercontent.com/175722/158036241-f3acf38f-60a2-463c-adc2-cd5c41af8983.png">

Example after, less than 2.5 minutes for Conda setup and Pandas build (both w/ hot cache)
<img width="93" alt="Screenshot 2022-03-12 at 22 46 50" src="https://user-images.githubusercontent.com/175722/158036242-90b30ce9-bbc1-43a8-b4fc-6ac3eaaba5a3.png">
